### ISSUE TITLE
Add a decoding flag to get all numbers as strings

### ISSUE THREAD

**Issue by akheron** [akheron/jansson#10]
All numbers or only overflowing numbers, or maybe two separate flags?

**Comment by keenerd** [comment]
For an example of how another library does this:  yajl-tree simply stores the original string and lets you access it directly: https://github.com/lloyd/yajl/blob/master/src/api/yajl_tree.h#L81

**Comment by akheron** [comment]
Yeah, this is also an option. But allocating the extra memory to store input strings of every number doesn't sound good for people using this on embedded devices. It could of course be enabled by a decoding flag to make it optional.

Would if be beneficial to do this also for strings? They have many possible input forms because any Unicode code point can be escaped with `\u` or represented directly in UTF-8.

**Comment by dbelliveau** [comment]
Because of the decimal precision issue we made a change to our copy of Jansson to allow all numbers to be treated as strings. This might not be efficient for many use cases but for ours it's actually more efficient. In brief:

* New flag JSON_DECODE_NUMBER_AS_STRING causes json_load* to decode all numbers as strings.
* New int field in json_string_t to indicate if a string is a number. Set to true when decoding if the source field is a number. When encoding, if true json_dump* will encode the string as a number (no quotes).
* New function json_string_is_number() returns true if a string is number.
* New function json_string_set_is_number() to set the "string is a number" indicator.

If there's interest in adding this to Jansson let me know and I'll create a pull request.

**Comment by nertpinx** [comment]
@dbelliveau If it's still possible to submit your patches as PR, then it might help with #425 (my guess is that it's taking similar approach but it comes from a different source).

**Comment by coreyfarrell** [comment]
@akheron any reason you are against creating a new `json_type` for this (maybe `JSON_RAW`)?  My complaint about the current patches is that it expands storage of json_string_t.  I realize adding a new type will require more care but I think it's worth the effort.  One thing I noticed about the patch @keenerd created is that JSON_SNUMBER is added to the middle of the enum, it would need to be last on the enum otherwise ABI would change.  Naming it JSON_RAW instead of JSON_SNUMBER would leave us the option to use that same type to support raw / non-decoded strings as well (possible micro-optimization for some use cases).

**Comment by akheron** [comment]
I'm not against adding a new type if it's not only for working around quirks of floating point formatting.

What are the use cases that we're trying to support here? I think I've lost track :)

Supporting numbers that don't have a native representation ("bignums") come to my mind, but are there other use cases?

**Comment by akheron** [comment]
Alright, sounds good.

Having a JSON_RAW value, how does the user know what it contains (number/string/etc)?

**Comment by coreyfarrell** [comment]
I was thinking that JSON_RAW would use json_string_t in the background and thus would not store further information about the contents.  In theory we could provide a function `json_type json_raw_typeof(json_t *json)` identify the contents by checking the first character of a JSON_RAW:
* `t` or `f` means it is a boolean
* `n` means it is a null
* `"` means it is a string
* `{` means it is an object
* `[` means it is an array
* `-0123456789` would be a number
* Anything else is an error

We would have json_raw support functions matching the json_string functions, the create/set functions would function slightly differently than json_string functions.  Create/set would strip white-space at the start / end but not within.
* json_raw() / json_raw_set would perform a validation only parse with JSON_DECODE_ANY enabled.
* _nocheck variants would not perform any validation but would still trim white-space.

The following code would print `strcmp: 0`:
```c
json_t *raw = json_raw(" \n\t[ 1, 2\n\t ] \n\t");
printf("strcmp: %d\n", strcmp(json_raw_value(raw), "[ 1, 2\n\t ]"));
```

**Comment by akheron** [comment]
Would it be useful to have an option to not store anything at all in JSON_RAW? That would make the JSON structure non-encodable, but on the other hand it would save memory if only a part of the input is interesting for the userm

**Comment by coreyfarrell** [comment]
I hadn't thought of that, I'm not sure.  My idea was that JSON_RAW would be used where the value is not locally interesting, but still needs to be saved to disk or forwarded to clients / other servers for processing.

**Comment by coreyfarrell** [comment]
@akheron I've done some work on an actual implementation of the JSON_RAW type, I think that the correct way to determine what a `JSON_RAW` contains would be to parse it with `JSON_DECODE_ANY` and look at the result.  In some cases you could make assumptions based on the code used to create the object (you might know it's only possible for certain types to be within the JSON_RAW fields).

I've posted #446 with the first attempt implementation of JSON_RAW.

**Comment by AllenX2018** [comment]
Is there any progress on this issue? @coreyfarrell Are you still working on the new json raw type?

**Comment by coreyfarrell** [comment]
#446 is the last I've worked on this.  Unfortunately I've been busy with other commitments lately.  Until I get some critical feedback on my PR it's unlikely to see any further progress.
